### DISCUSSION TITLE
qpdf external dependencies

### DISCUSSION THREAD

**Discussion by m-holger** [qpdf/qpdf-dev#14] [General]
From https://github.com/qpdf/qpdf/pull/1274#issuecomment-2312704015 :

> I have often considered whether qpdf should split into a core library with minimal dependencies (perhaps just zlib and openssl/gnutls) and one or more libraries layered on top. A big part of qpdf's success has been that it works and builds anywhere. Then an outer library could add support for the image filters, xml metadata, and maybe some other things.

I am not sure what exactly you have in mind. I certainly think that there is a good argument for making greater use of external libraries and only providing some functionality when the necessary libraries are available. Taking  the Info / Metadata as an example, I think it is entirely reasonable to only provide support for moving info to metadata if a suitable XML library is available, and provide support for dumping info to eg JSON and importing user formatted metadata from a text file / string.

**Comment by jberkenbilt** [comment]
Use of external libraries could also potentially enable working with fonts, JBIG2 or JP2 compression, or potentially g3 and g4 compression though those are trickier because the PDF way of doing it is a little different from the tiff way of doing it. libtiff neither exposes APIs for that nor implements the full range of what PDF is needed. Anyway, just musing. I think other things are higher priority. Only one person has ever asked about CCITT group 4 compression support and no one has ever asked about the other image formats, where the pages epic stuff comes up all the time.

**Reply by m-holger** [reply]
I believe pikepdf has some support for CCITT group 4 encoding, so there may be a demand that is supported elsewhere.

**Comment by m-holger** [comment]
I think another area where external libraries might be useful are more general utilities more peripheral to qpdf's core purpose, such as command line parsing, terminal io and info/error message formatting. Making use of external libraries here has the potential benefit of freeing up effort for more core activities while potentially providing more powerful solutions as well as more standard solutions. 

The problem with this is of course that C++ does not have a comprehensive standard library like eg Python, and therefore one cannot rely  on the required libraries being available everywhere. For this to fly I think one of two things have to be true:
- the library is header only
- the library is widely available and a fallback for when it is not is relatively trivial to implement. I am thinking here for example of GNU  gettext for warning / error messages.
